regexp/regexp.go

```diff
@@ -0,0 +1,98 @@
+package regexp
+
+import (
+	"regexp"
+	"runtime"
+	"sync"
+	"weak"
+)
+
+// Interns the compilation of Regular Expressions.  If two regexs with the same pattern are compiled, the result
+// is the same *regexp.Regexp.  This avoids the compilation cost but more importantly the memory usage.
+//
+// Regular expressions produced from this package are backed by a form of weak-valued map, upon a regexp becoming
+// unreachable, they will be eventually removed from the map and memory reclaimed.
+
+var (
+	weakMap      = make(map[string]weak.Pointer[regexp.Regexp])
+	posixWeakMap = make(map[string]weak.Pointer[regexp.Regexp])
+	reverseMap   = make(map[weak.Pointer[regexp.Regexp]]string)
+	l            sync.Mutex
+)
+
+// CompileInterned compiles and interns a regular expression and returns a
+// pointer to it or an error.
+func CompileInterned(pattern string) (*regexp.Regexp, error) {
+	return compile(pattern, regexp.Compile, weakMap)
+}
+
+// CompilePOSIXInterned compiles and interns a regular expression using POSIX
+// syntax and returns a pointer to it or an error.
+func CompilePOSIXInterned(pattern string) (*regexp.Regexp, error) {
+	return compile(pattern, regexp.CompilePOSIX, posixWeakMap)
+}
+
+// MustCompileInterned compiles and interns a regular expression and returns a pointer to it or panics.
+func MustCompileInterned(pattern string) *regexp.Regexp {
+	return mustCompile(pattern, regexp.MustCompile, weakMap)
+}
+
+// MustCompilePOSIXInterned compiles and interns a regular expression using
+// POSIX syntax and returns a pointer to it or panics.
+func MustCompilePOSIXInterned(pattern string) *regexp.Regexp {
+	return mustCompile(pattern, regexp.MustCompilePOSIX, posixWeakMap)
+}
+
+// compile handles compiling and interning regular expressions. If the regexp is
+// already interned, a pointer to it is returned from the map. If the regexp is
+// not interned or is nil (since it's a weak pointer), it is compiled and stored
+// in the maps. The regexp is stored in the maps as a weak pointer, so that it
+// can be garbage collected.
+func compile(pattern string, compileFunc func(string) (*regexp.Regexp, error), m map[string]weak.Pointer[regexp.Regexp]) (*regexp.Regexp, error) {
+	l.Lock()
+	defer l.Unlock()
+	if itemPtr, ok := m[pattern]; ok {
+		ptr := itemPtr.Value()
+		if ptr != nil {
+			return ptr, nil
+		}
+		delete(m, pattern)
+		delete(posixWeakMap, pattern)
```

Comment:
I'm curious about the pattern here where we only insert into whatever map we're called with, `m`, but we delete from both `m` and `posixWeakMap`. It seems like there's a subtle behavior change here, where when `m` is `weakMap`, we're deleting from both maps, but when `m` is `posixWeakMap`, we're only deleting from `posixWeakMap` and not from `weakMap`. I wonder if it would be better to just delete from `m`?

Reply:
Sorry for confusion, you caught me mid-refactor here and I should have toggled the PR to draft or left a comment.

You are spot on here, Bruno also pointed out this issue in a thread above and we think `cleanup()` might also need to be changed, but I didn't address these issues yet. I will tag you in the other thread, so that we can carry on the conversation there.

<img src="https://media2.giphy.com/media/3o7WIxPr8EvjpDlfws/giphy.gif?cid=bd3ea57ebdd7qvu4543lu98vkcyr2ecdvwqho8m1anxnqd42&ep=v1_gifs_search&rid=giphy.gif&ct=g"/>

Reply:
Fixed in b21b980